From 794eb84a9b5a8d0b07b906a8c3af1c142a83ef47 Mon Sep 17 00:00:00 2001
From: Vivek Goyal <vgoyal@redhat.com>
Date: Mon, 22 Sep 2014 19:41:35 -0400
Subject: [security] keys: Reinstate EPERM for a key type name beginning with a '.'

Message-id: <1411414897-26220-3-git-send-email-vgoyal@redhat.com>
Patchwork-id: 96448
O-Subject: [RHEL7.1 PATCH 2/4] KEYS: Reinstate EPERM for a key type name beginning with a '.'
Bugzilla: 1129779
RH-Acked-by: David Howells <dhowells@redhat.com>

BZ:1129779

This commit id is in linux-security tree of James Morris.

commit 54e2c2c1a9d6cbb270b0999a38545fa9a69bee43
Author: David Howells <dhowells@redhat.com>
Date: Tue, 16 Sep 2014 17:29:03 +0100

Reinstate the generation of EPERM for a key type name beginning with a '.' in
a userspace call.  Types whose name begins with a '.' are internal only.

The test was removed by:

 commit a4e3b8d79a5c6d40f4a9703abf7fe3abcc6c3b8d
 Author: Mimi Zohar <zohar@linux.vnet.ibm.com>
 Date:   Thu May 22 14:02:23 2014 -0400
 Subject: KEYS: special dot prefixed keyring name bug fix

I think we want to keep the restriction on type name so that userspace can't
add keys of a special internal type.

Note that removal of the test causes several of the tests in the keyutils
testsuite to fail.

Signed-off-by: David Howells <dhowells@redhat.com>
Acked-by: Vivek Goyal <vgoyal@redhat.com>
cc: Mimi Zohar <zohar@linux.vnet.ibm.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/security/keys/keyctl.c b/security/keys/keyctl.c
index da50c93..762cb3d 100644
--- a/security/keys/keyctl.c
+++ b/security/keys/keyctl.c
@@ -37,6 +37,8 @@ static int key_get_type_from_user(char *type,
   return ret;
  if (ret == 0 || ret >= len)
   return -EINVAL;
+ if (type[0] == '.')
+  return -EPERM;
  type[len - 1] = '\0';
  return 0;
 }
-- 
1.7.1